File: PHP/References/TSub.php
<html>
   <head>
      <title>PHP Test Subprograms </title>
   </head>
   <body>
      <?php
         // Test Subprograms: Php Call by Value and Call by Reference
         // Program-ID: TSub.php
         // Author:     Kuo-pao Yang
         // OS:         Ubuntu 24
         // Parser:     PHP 8

         //Test call by value 
         print "PHP: Test Call by Value<br>";
         $m = 1; 
         func1($m);
         print "m = ".$m."<br>";
         $n = func2($m);
         print "n = ".$n."<br>";
         //Test Array to Subprogram (Call-By-Value)
         print "Test Array to Subprogram (Call-By-Value)<br>";
         $a = array(10, 20, 30); 
         func3($a[1], $a);
         for($i = 0; $i < 3; $i++) {
            print "a[".$i."] = ".$a[$i]." ";
         }
         print "<br>";
         $b = func4($a[1], $a);
         for($i = 0; $i < 3; $i++) {
            print "b[".$i."] = ".$b[$i]." ";
         }
         print "<br>";

         //Test Array to Subprogram (Call-By-Reference)
         print "Test Array to Subprogram (Call-By-Reference)<br>";
         $c = array(10, 20, 30); 
         func5($c[1], $c);
         for($i = 0; $i < 3; $i++) {
            print "c[".$i."] = ".$c[$i]." ";
         }

         function func1($i) { 
            $i = $i + 1; 
         }
         function func2($i) { 
            $i = $i + 2; 
            return $i; 
         }
         function func3($i, $j) { 
            $i = $i + 3; $j[0] = $j[1] + 4; 
         }
         function func4($i, $j) { 
            $i = $i + 3; $j[0] = $j[1] + 4; 
            return $j;
         }
         function func5(&$i, &$j) { 
            $i = $i + 3; $j[0] = $j[1] + 4; 
         }

         /* Output:
         PHP: Test Call by Value
         m = 1
         n = 3
         Test Array to Subprogram (Call-By-Value)
         a[0] = 10 a[1] = 20 a[2] = 30
         b[0] = 24 b[1] = 20 b[2] = 30
         Test Array to Subprogram (Call-By-Reference)
         c[0] = 27 c[1] = 23 c[2] = 30 
         */
      ?>
   </body>
</html>

File: Progamming_Assignments/C/Sample_Code/TSub.c
// Test Subprograms: Call by Value and Call by Reference
// Program-ID:   TSub.c
// Author:       Kuo-pao Yang
// OS:           Ubuntu 24
// Compiler:     GNU C
// Note:
// The following instructions are used to
//      edit, compile, and run this program
//    $ nano     TSub.c
//    $ gcc      TSub.c
//    $ ./a.out

#include <stdio.h>

int func1(int i1, int *j1) {
    i1 = i1 + 1;
    *j1 = *j1 + 2;
    printf("func1() i1 = %d, j1 = %d\n", i1, *j1);
    return (i1 + *j1);
}
void func2(int i2, int j2[]) {
    i2 = i2 + 3;
    j2[0] = j2[1] + 4;
    printf("func2() i2 = %d, j2 = %d\n", i2, j2[0]);
}
void func3(int i3, int *j3){
    i3 = i3 + 3;
    *j3 = *(j3 + 1) + 4;
    printf("func3() i3 = %d, j3 = %d\n", i3, *j3);
}

void main() {
    //Test call by value and call by reference
    printf("Test call by value and call by reference\n");
    int n1 = 1, n2 = 2;
    int n3 = func1(n1, &n2);
    printf("n1 = %d, n2 = %d, n3 = %d\n", n1, n2, n3);

    //Test Array to Subprogram
    int i;
    int a[3] = {10, 20, 30}; 
    printf("Test Array to Subprogram: way 1 (array)\n");
    func2(a[1], a);
    for(i = 0; i < 3; i++) {
       printf("a[%d] = %d\t", i, a[i]);
    }
    printf("\nTest Array to Subprogram: way 2 (pointer, same result)\n");
    a[0] = 10; a[1] = 20; a[2] = 30;
    func3(a[1], a);
    for(i = 0; i < 3; i++) {
       printf("a[%d] = %d\t", i, a[i]);
    }
}

/* Output:
Test call by value and call by reference
func1() i1 = 2, j1 = 4
n1 = 1, n2 = 4, n3 = 6
Test Array to Subprogram: way 1 (array)
func2() i2 = 23, j2 = 24
a[0] = 24       a[1] = 20       a[2] = 30
Test Array to Subprogram: way 2 (pointer, same result)
func3() i3 = 23, j3 = 24
a[0] = 24       a[1] = 20       a[2] = 30           
*/
